crates/red_knot_project/src/metadata/options.rs

```
@@ -113,6 +113,7 @@ impl Options {
                         .ok()
                         .map(PythonPath::from_virtual_env_var)
                 })
+                .or_else(PythonPath::from_local_venv)
```

I'm not sure if implementing the fallback here, at least how it's done now is the most ideal solution. I've two concerns:

1. This re-implements part of the discovery logic in `VirtualEnvironment::new`
2. It's now possible that Red Knot picks up a `.venv` for which `VirtualEnvironment::new` later fails (for example, because the `pyenv.cfg` is invalid). Red Knot will fail to start

I think my preferred behavior is that Red Knot ignores any errors (may log a warning or debug message) if a `.venv` directory exists but it isn't a valid venv. But I don't think Red Knot should fail to start in that case because it isn't an input explicitly set by the user.

One way to implement this is to have a `PythonPath::Discover` option where the fallback behavior is implemented in `SearchPaths::from_settings`

---

> It's now possible that Red Knot picks up a .venv for which VirtualEnvironment::new later fails (for example, because the pyenv.cfg is invalid). Red Knot will fail to start

Could this not happen already?

---

Could we say we just don't error for site packages discovery for?
```rs
SysPrefixPathOrigin::VirtualEnvVar | SysPrefixPathOrigin::LocalVenv
```
?

---

> Could this not happen already? It could pick up the env variable of an invalid .venv right?

IMO, this is different because the user explicitly sets the `VIRTUAL_ENV` environment variable. They're explicitly telling Red Knot to use that folder.

> Could we say we just don't error for site packages discovery for these?

We could do that. But I wonder if it still leads to some more code duplication compared to deferring more to the resolver.

---

Do you think its still bad to look for a venv there
```rs
                .or_else(PythonPath::from_local_venv)
```
Then in SearchPaths::from_settings, we allow an invalid venv, and fallback to no site packages?

---

I don't think it's bad in itself. It even has some advantages from a persistent caching perspective (long-term) because we'd calculate a different cache key when running red knot when the `.venv` folder exists vs when it isn't. 

However, I do wonder if we could remove some code duplication (the code roughly repeats some of the discovery logic that we already have in resolver) by moving this logic entirely into the resolver. 

There's a last part to this feature, but I think it's okay to tackle it separately: We need to re-run the event discovery if the `.venv` folder gets deleted or created.

---

No, I'd introduce a new `PythonPath` variant because the resolver otherwise can't know that it should do the venv discovery (`KnownSitePackages` would then become testing only)

---

Okay I think understand, I'll have a go at this

---

How does that look?

---

Sorry yeah now `KnownSitePackages` is only used in tests